# Pandas - Data Correlations

## Finding Relationships

A great aspect of the Pandas module is the `corr()` method.

The `corr()` method calculates the relationship between each column in your data set.

We will use `data.csv` dataset:

In [1]:
import pandas as pd

df = pd.read_csv('data.csv')

print(df)

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]


Eg: Show the relationship between the columns:

In [2]:
df.corr()

Duration     Pulse  Maxpulse  Calories
Duration  1.000000 -0.155408  0.009403  0.922717
Pulse    -0.155408  1.000000  0.786535  0.025121
Maxpulse  0.009403  0.786535  1.000000  0.203813
Calories  0.922717  0.025121  0.203813  1.000000

> **Note**: The `corr()` method ignores "not numeric" columns.

## Result Expained

The Result of the `corr()` method is a table with a lot of no.s that represents how well the relationship is between two columns.<br><br><br>

The number varies from -1 to 1.<br><br>   

1 means that there is a 1 to 1 relationship (a perfect correlation), and for this data set, each time a value went up in the first column, the other one went up as well.<br><br>

0.9 is also a good relationship, and if you increase one value, the other will probably increase as well.<br><br>

-0.9 would be just as good relationship as 0.9, but if you increase one value, the other will probably go down.<br><br>

0.2 means NOT a good relationship, meaning that if one values goes up does not mean that the other will.<br><br><br>

> **What is a good correlation?** It depends on the use, but I think it is safe to say you have to have at least `0.6` (or `-0.6`) to call it a good correlation.

Perfect Correlation:

We can see that "Duration" and "Duration" got the no. `1.000000`, which makes sense, each column always has a perfect relationship with itself.

## Good Correlation:

"Duration" and "Calories" got a `0.922721` correlation, which is a very good correlation, and we can predict that the longer you work out, the more caloreis you burn, and the other way around: if you burned a lot of calories, you probably had a long work out.

## Bad Correlation:

"Duration" and "Maxpulse" got a `0.009403` correlation, which is a very bad correlation, meaning that we can not predict the max pulse by just looking at the duration of the work out, and vice versa.